### DISCUSSION TITLE
Improvements to the UI

### DISCUSSION THREAD

**Discussion by LiamD-Flop** [YimMenu/YimMenu#866] [Ideas]
#865 
As stated above. Tell us what is wrong with the current UI / give us ideas on how we could improve it.

**Comment by ghost** [comment]
One of the things about this menu I'm not a big fan of is that it captures the mouse, but I can see with how it currently is built that that would be hard to do without. Not a fan but I get it. I'd much rather prefer everything was navigable with the 10-key, like how kiddions and stand are. 

Another thing I'm no fan of, and yimmenu is no exception here, is how buried in menu navigation everything is, especially settings. There's probably no easy way to fix that but consider a check-all kind of thing for the reactions section, or check-all for logging, for example. 

An option to hide items would be nice, too. Especially for things I'll never use. I don't care about troll options, or flying cars, or god mode, or rainbow cars, for example. So an option to hide menu items one doesn't use could be nice for cleaning up the menu and making it easier to use while online. 

It would be nice if it was designed to take up as little screen real estate as possible. Stand and kiddions both do this well, as an example. Yimmenu appears to have no qualms about blocking the whole screen if desired. 

The notifications in the top right section of the screen do not appear to get logged and don't have a consistent look as the rest of the menu. Yet yimmenu also makes use of the game's actual notification section at the bottom left (such as with player join notifications, which I love this feature, btw, and wish it was built into the game). In my opinion the look and feel needs to be unified. One notification area would be nice, too. I'd prefer the bottom left like just about everything else uses. I'll consistently miss the top right ones, so having them logged would be nice, too. 

Speaking of logs, the player database file is not human-readable. It saves everything to a single line and is hard to parse through. I like the search function in the player database but wish it rather had an option to hide all not currently in the session. If it showed only ~32 entries then I suppose we'd never have to search or scroll too unless viewing the whole list. If there was some way to combine the player database functionality and the session player list functionality that would be nice, too. This doesn't feel like it should be two different sections. 

That's the main feedback I have on using yimmenu's UI. 

EDIT: with regard to logging, it'd be nice if repeated items had a "repeated X number of times" or "x400" instead of actually logging the same info on 400 lines of the log).

**Comment by tupoy-ya** [comment]
### What is wrong with the current UI

- Screen space
  Current UI takes a lot of screen space and with a game like GTA Online where screen space is already in short supply having huge windows isn't great.
- Customization
  With the current UI there is no customization at all, so you can't change the default placement or make windows smaller or hide something you don't need.
- Default tabs
  When you select Ped or Vehicle menu you will get into them directly without selecting a submenu but doing the same with other menus doesn't work.

There is also the awful log toggle placement where some toggles are in reactions tab and some in debug tab and some in session tab.


### About keyboard navigation
About keyboard navigation: It's possible just hard to use with default ImGui widgets.
We can also make so that holding `C` or some other key will get us mouse input after the GUI is open.

Here is an example of keyboard navigation in my object spooner test:

https://user-images.githubusercontent.com/72797377/212696765-87b8842f-aa9d-4fec-8750-3e87fc58ee0d.mp4

**Comment by xiaoxiao921** [comment]
Bumping this as it's getting almost critical in certain areas, like when you select another player
The toxic UI tree is a total mess, hard to believe it's like this currently
Also teleport to player under "troll" which makes no sense

**Reply by Seanghost117** [reply]
Agreed there needs to be either more sub division or consistency. Something akin to Nice, Troll, Toxic, Vehicles, Misc or along the lines. Feels like a lot just gets thrown into one out of ease even if it doesn't make sense.

**Comment by Zuwy101** [comment]
please dont make it keyboard navigation, keep in mouse mode. I strongly dislike keyboard navigation.

**Reply by ghost** [reply]
if it ever would become kb navigation it would be nice to atleast have an option to switch from mouse to kb and from kb to mouse controls

**Comment by Seanghost117** [comment]
**My take / Current Suggestion**
Areas to Improve:

- Right now the UI is a mess, things are thrown places, it's not properly handled and some areas are too general
- There are a lot of buttons, especially where the toxic node tree is, this could be condensed to more combos or another method because right now it's just getting spammy
- The UI is able to fill the screen with no constraint, as more features have been added, and some tabs are more than literally taking up a vast majority of the screen.
- The size of elements and text themself is also decently big, it's made for people over the age of 40 with bad eyesight, another area that can be condensed.
- There is minimal customization, besides changing the color of the windows background which is really where it ends.
- Management of tabs, some areas could benefit from a misc tab, a lot of sub-tabs in the world can just be put into a main tab as the world tab doesn't have a default view, and sections like water could be removed in favor of this or expanded.

My Suggestions:

- First let's clean up the player tab, instead of node trees we may want to prefer vertical tabs, either using buttons or a custom tab as tab bars are kinda ugly imo
- We need a hard constraint on the window, this means either hard setting the size or making sure there is an x and y constraint in effect
- More elements can benefit from color options, especially basic imgui elements such as buttons, button text, child backgrounds if they are used, etc.
- There needs to be an implementation of a style manager, this isn't hard we just need to port the color and ui variables to a standalone theme manager and have the selected theme be updated in the default config for saving and loading, and sharing
- although there are perks to having the player tab always drawn under the main tabs we can also discuss having a standalone portion in a window or just throwing it in a list box, especially with the size of the elements (being made for 40 years olds) again its bulky and spammy.
- It's a imgui menu running on dx, we can add our own personal flairs, integrate image support, etc, things like this we can think outside the box
- More general tabs need to be with more general tabs, weapons could benefit from being standalone and aimbot as a sub-tab instead of putting it in self.
- The tooltip can be reworked to be in a more friendly position and have more of a flare to it / as well as include the command to use it instead of just the description.
- A command lookup feature can be helpful by looking up the command by its friendly name
- We can also add side views for player information, vehicle details for the spawner, etc

It feels like after the initial draft of the current UI the UI itself was just abandoned, and just as we do feature updates we should be expanding the ui and customizability to give more control to the user. These are my honest takes and opinion, feel free to reply with questions or points that I brought up

**Reply by LiamD-Flop** [reply]
I would like to add onto this. As most of the issues with the current ui are caused by a lot of new features being added to YimMenu but there being no supervision on where or how they are put into the menu. Currently the testing of pr are done by just testing the feature, regardless of how to interact with it. I would like to suggest adding this to the pr approval steps. Doing this there will be some management before it gets out of hand again. Therefor we "hopefully" would not need a discussion like this in the future.

**Comment by Tayipc38** [comment]
Right now I have made 3 major overlay changes. First being the options modal:
![image](https://github.com/YimMenu/YimMenu/assets/79384354/e5a71766-f340-4188-b3e8-0627962ec307)
![image](https://github.com/YimMenu/YimMenu/assets/79384354/65143d0c-31d8-4759-a40d-7d7ea69efaa0)

This modal is meant to be used for UI elements that aren't used heavily, such as feature settings or extra abilities.
The second change was having features separated to ListBoxes:
![image](https://github.com/YimMenu/YimMenu/assets/79384354/05fcec3e-8441-465d-8f59-48fea5065e95)
Although it helps organize a little bit, some portions are so overflooded that many features are initially locked behind a scrollbar. 

The last change I made was merging similar tabs into one:
![image](https://github.com/YimMenu/YimMenu/assets/79384354/60796472-4578-420f-90c6-36464e2a49b0)
Personally it improved the times it took to reach some of these, but I can sense some issues with clarity.

If anyone has a fix, improvement or total dislike of these changes, feel free to sound your opinion,

**Reply by Tayipc38** [reply]
And I am a recent fan of `ImGui::SeparatorText` which I will be integrating in future PR's.
